File: experimental/forthic-cpp/ForthicLib/Words/EndArrayWord.cpp
#include "pch.h"
#include <algorithm>
#include "EndArrayWord.h"
#include "../Interpreter.h"
#include "../StackItems/ArrayItem.h"
#include "../StackItems/StartArrayItem.h"


EndArrayWord::EndArrayWord(string word_name) : Word(word_name)
{
}


EndArrayWord::~EndArrayWord()
{
}

void EndArrayWord::Execute(Interpreter *interp)
{
	vector<shared_ptr<StackItem>> result;

	while (true)
	{
		auto item = interp->StackPop();
		if (dynamic_cast<StartArrayItem*>(item.get())) break;
		else result.push_back(item);
	}

	std::reverse(result.begin(), result.end());
	interp->StackPush(shared_ptr<StackItem>(new ArrayItem(result)));
}

File: forthic-rb/lib/forthic/words/end_array_word.rb
# # frozen_string_literal: true

require_relative "word"
require_relative "../token"

module Forthic
  class EndArrayWord < Word
    def initialize
      super("]")
    end

    # @param [Interpreter] interp
    def execute(interp)
      items = []
      item = interp.stack_pop

      # NOTE: This won't infinite loop because interp.stack_pop() will eventually fail
      loop do
        break if item.is_a?(Token) && item.type == TokenType::START_ARRAY
        items.push(item)
        item = interp.stack_pop
      end

      items.reverse!
      interp.stack_push(items)
    end
  end
end
